keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
+
+The rules that govern these fields are described in the below sections
+
+##### Creating a New Single Stack Service
+
+User who want to create a new single stack Service can provide `nil` values for both `spec.ipFamilyPolicy` and `spec.ipFamilies`. The results of this are:
+1. the new Service will be defaulted to `spec.ipFamilyPolicy: SingleStack`.
+2. The new service will be defaulted to `spec.ipFamilies[<cluster default address family>]`.
+3. A single IP from that family will be allocated to that service.
+
+Example for a service that has been created using the above approach on a cluster that has `IPv4` as cluster default address family.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  ipFamilyPolicy: SingleStack # defaulted
+  ipFamilies: # defaulted
+    - IPv4
+  clusterIP: 1.2.3.4 # allocated
+  clusterIPs:
+    - 1.2.3.4 
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+```
+
+> the above is also used to `pin` existing cluster services (pre convert-to-dualstack action) to their respective families.
+
+Users might be also in creating a single stack service but allocated from the alternative address family. E.g. Creating `IPv6` single stack Service for a cluster that has `--services-cluster-ip-range=<ipv4 cidr>, <ipv6 cidr>`. For this scenario, users can set `spec.ipFamilies[ 'IPv6' ]` in Service spec. The results of this are:
+1. The new Service will be defaulted to `spec.ipFamilyPolicy: SingleStack`.
+2.  A single IP from that family will be allocated to that service following the family requested by user.
+
+Example for a service that has been created using the above approach.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  ipFamilyPolicy: SingleStack # defaulted
+  ipFamilies: # set by user
+    - IPv6
+  clusterIP: 2001::1 # allocated
+  clusterIPs:
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+```
 
-For example, the following Service specification includes the `spec.ipFamily` field. Kubernetes will assign an IPv6 address (also known as a “cluster IP”) from the configured service-cluster-ip-range to this Service.
+> In single stack services users are free to set `spec.ipFamilyPolicy` and `spec.ipFamilies`. The system will validate that `spec.ipFamilyPolicy: SingleStack` case have a matching single entry in `spec.ipFamilies[]`
+
+##### Creating a New Dual Stack Service
+
+Users can create dual stack services and control which up gets assigned to Service Primary IP as the following:
+1. Set `spec.ipFamilyPolicy: DualStack`. 
+2. If `spec.ipFamilyPolicy: nil` and `len(spec.ipFamilies) == 2` the resulting service will have a defaulted `spec.ipFamilyPolicy: DualStack`.
+3. If `spec.ipFamilies: nil` then it will be defaulted to the same order of ip families as set in `--services-cluster-ip-range`
```

Comment:
@danwinship 's comments make me question this.  I thought it was good that Services get `policy` hard-pinned like this, but I am less than 100%.

Thinking thru cases...

* If I specify `ipFamilies = [ v4 ]; ipfamilyPoilcy = nil`, would I want admission control to change that to DualStack?  I guess not.  I mean, I could make an argument that this expressing "at least v4", but that's a stretch.

OK, I talked myself out of it.